### PR TITLE
Add delete method to cache

### PR DESCRIPTION
**Pull request by maggie-lou**
Splitting up this PR: https://github.com/buildbuddy-io/buildbuddy/pull/2288

### CODE REVIEW

**Comment by maggie-lou** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2349** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/2349" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/2349?utm_source=stack-comment).

File: enterprise/server/backends/distributed/distributed_test.go [hunk 1]
```diff
@@ -846,3 +846,104 @@ func TestHintedHandoff(t *testing.T) {
...
+		memoryCache3,
+	}
+	distributedCaches := []interfaces.Cache{dc1, dc2, dc3}
+
+	for i := 0; i < 100; i++ {
+		// Do a write, which should be written to all nodes.
+		d, buf := testdigest.NewRandomDigestBuf(t, 100)
+		if err := distributedCaches[i%3].Set(ctx, d, buf); err != nil {
+			t.Fatal(err)
+		}
```
Comment (bduffany) [inline enterprise/server/backends/distributed/distributed_test.go:894]: nit: add a Contains() test after Setting ?

File: enterprise/server/backends/distributed/distributed_test.go [hunk 2]
```diff
@@ -846,3 +846,104 @@ func TestHintedHandoff(t *testing.T) {
...
+	waitForReady(t, config3.ListenAddr)
+
+	distributedCaches := []interfaces.Cache{dc1, dc2, dc3}
+
+	for i := 0; i < 100; i++ {
+		d, _ := testdigest.NewRandomDigestBuf(t, 100)
+
+		// Do a delete on a file that does not exist.
+		err := distributedCaches[i%3].Delete(ctx, d)
+		assert.Nil(t, err)
```
Comment (bduffany) [inline enterprise/server/backends/distributed/distributed_test.go:947]:
```suggestion
		assert.NoError(t, err)
```

File: enterprise/server/util/cacheproxy/cacheproxy_test.go [hunk 3]
```diff
@@ -777,3 +777,42 @@ func TestEmptyRead(t *testing.T) {
...
 
 }
+
+func TestDelete(t *testing.T) {
+	ctx := context.Background()
+	te := getTestEnv(t, emptyUserMap)
+
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, te)
+	if err != nil {
+		t.Errorf("error attaching user prefix: %v", err)
```
Comment (bduffany) [inline enterprise/server/util/cacheproxy/cacheproxy_test.go:787]:
```suggestion
		t.Fatalf("error attaching user prefix: %v", err)
```

**Review by bduffany — COMMENTED** [review]
will let Tyler or Vadim approve

File: enterprise/server/backends/distributed/distributed_test.go (hunk 1 above)
Comment (maggie-lou) [reply enterprise/server/backends/distributed/distributed_test.go:894]: Done

File: enterprise/server/backends/distributed/distributed_test.go (hunk 2 above)
Comment (maggie-lou) [reply enterprise/server/backends/distributed/distributed_test.go]: Done

File: enterprise/server/util/cacheproxy/cacheproxy_test.go (hunk 3 above)
Comment (maggie-lou) [reply enterprise/server/util/cacheproxy/cacheproxy_test.go]: Done

File: enterprise/server/backends/distributed/distributed.go [hunk 4]
```diff
@@ -973,7 +979,18 @@ func (c *Cache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error
 }
 
 func (c *Cache) Delete(ctx context.Context, d *repb.Digest) error {
-	return status.UnimplementedError("Not yet implemented.")
+	ps := c.readPeers(d)
+	for peer := ps.GetNextPeer(); peer != ""; peer = ps.GetNextPeer() {
+		err := c.remoteDelete(ctx, peer, c.isolation, d)
+		if err != nil {
+			if status.IsNotFoundError(err) {
+				c.log.Warningf("File not found, could not be deleted: %s", err.Error())
```
Comment (vadimberezniker) [inline enterprise/server/backends/distributed/distributed.go:987]: Do we care if it doesn't exist? I'm not sure it's worth a warning.

**Review by vadimberezniker — APPROVED** [review]

File: enterprise/server/backends/distributed/distributed.go (hunk 4 above)
Comment (maggie-lou) [reply enterprise/server/backends/distributed/distributed.go]: I thought a warning might be nice if you had a typo in the file name or something, but the warning logic isn't consistent so I removed it (i.e. if the data wasn't written to a peer as expected, client shouldn't be warned about that. Also some of the cache types, like S3, don't expose whether the file existed or not before deletion)

**Pull request merged**